retranmission.go

```diff
@@ -0,0 +1,203 @@
+	// to other side.
+	messagesToRetransmit []lnwire.Message
+
+	mutex sync.RWMutex
+}
+
+// newRetransmitter creates new instance of retransmitter.
+func newRetransmitter(storage channeldb.MessageStore) (*retransmitter, error) {
+
+	// Retrieve the messages from the message storage with their
+	// associated indexes.
+	indexes, messages, err := storage.Get()
+	if err != channeldb.ErrPeerMessagesNotFound && err != nil {
+		return nil, err
+	}
+
+	// Initialize map of code to index map, so than later we can retrieve
+	// indexes of messages that should be removed.
+	codeToIndex := make(map[lnwire.MessageCode][]uint64)
+	for i, message := range messages {
+		codeToIndex[message.Command()] = append(
+			codeToIndex[message.Command()],
+			indexes[i],
+		)
+	}
+
+	return &retransmitter{
+		storage:              storage,
+		codeToIndex:          codeToIndex,
+		messagesToRetransmit: messages,
+	}, nil
+}
+
+// Register adds message that should be acknowledged in the message storage.
+func (rt *retransmitter) Register(msg lnwire.Message) error {
+	switch msg.Command() {
+	// messages without acknowledgment
+	case lnwire.CmdCloseComplete,
+		lnwire.CmdChannelUpdateAnnouncement,
+		lnwire.CmdChannelAnnouncement,
+		lnwire.CmdNodeAnnouncement,
+		lnwire.CmdPing,
+		lnwire.CmdPong,
+		lnwire.CmdErrorGeneric,
+		lnwire.CmdInit:
+		return nil
+	default:
+		// Adds message to storage and returns the message index which
+		// have been associated with this message within the storage.
+		index, err := rt.storage.Add(msg)
+		if err != nil {
+			return err
+		}
+
+		// Associate the message index within the message storage
+		// with message code in order to remove messages by index later.
+		rt.mutex.Lock()
+		rt.codeToIndex[msg.Command()] = append(
+			rt.codeToIndex[msg.Command()], index,
+		)
+		rt.mutex.Unlock()
+
+		return nil
+	}
+}
+
+// Ack encapsulates the specification logic about which messages should be
+// acknowledged by receiving this one.
+func (rt *retransmitter) Ack(msg lnwire.Message) error {
+	switch msg.Command() {
+
+	case lnwire.CmdSingleFundingResponse:
+		return rt.remove(
+			lnwire.CmdSingleFundingRequest,
+		)
+	case lnwire.CmdSingleFundingComplete:
+		return rt.remove(
+			lnwire.CmdSingleFundingResponse,
+		)
+	case lnwire.CmdSingleFundingSignComplete:
+		return rt.remove(
+			lnwire.CmdSingleFundingComplete,
+		)
+	case lnwire.CmdFundingLocked:
+		return rt.remove(
+			lnwire.CmdSingleFundingSignComplete,
+		)
+	case lnwire.CmdUpdateAddHTLC,
+		lnwire.CmdUpdateFailHTLC,
+		lnwire.CmdUpdateFufillHTLC:
+		return rt.remove(
+			lnwire.CmdFundingLocked,
+		)
+	case lnwire.CmdRevokeAndAck:
+		return rt.remove(
+			lnwire.CmdUpdateAddHTLC,
+			lnwire.CmdUpdateFailHTLC,
+			lnwire.CmdUpdateFufillHTLC,
+			lnwire.CmdCommitSig,
+			lnwire.CmdCloseRequest,
+			lnwire.CmdFundingLocked,
+		)
+
+	case lnwire.CmdCommitSig,
+		lnwire.CmdCloseRequest:
+		return rt.remove(
+			lnwire.CmdFundingLocked,
+			lnwire.CmdRevokeAndAck,
```

Comment:
For now, all instances of `RevokeAndAck` should be omitted from retransmission. As is now, because we still use an "initial revocation window" of 1, peer restarts will cause `lnd` to send the initial `RevokedAndAck` _twice_ with the same revocation values. This'll cause the channel to fail down the line as a state transition will re-use the same preimage rather than going to the next leaf node in the tree.